File: samples/cext-trey-research-python/db_setup/azure_table_setup.py

import os
import sys
import json
import time
import uuid
from azure.data.tables import TableServiceClient, TableClient
from pathlib import Path

TABLE_NAMES = ["Project", "Consultant", "Assignment"]

def get_tables(table_service_client):
    tables = []
    table_iter = table_service_client.list_tables()
    for table in table_iter:
        tables.append(table.name)
    return tables

def delete_tables(table_service_client, connection_string):
    tables = get_tables(table_service_client)
    for table in tables:
        table_client = TableClient.from_connection_string(conn_str=connection_string, table_name=table)
        print(f"Deleting table: {table}")
        table_client.delete_table()
    
    while True:
        print("Waiting for tables to be deleted...")
        tables = get_tables(table_service_client)
        if not tables:
            print("All tables deleted.")
            break
        time.sleep(1)

def create_and_populate_tables(table_service_client, connection_string):
    for table_name in TABLE_NAMES:
        tables = get_tables(table_service_client)
        if table_name in tables:
            print(f"Table {table_name} already exists, skipping...")
            continue

        print(f"Creating table: {table_name}")
        table_created = False
        while not table_created:
            try:
                table_service_client.create_table(table_name=table_name)
                table_created = True
            except Exception as e:
                if '409' in str(e):
                    print('Table is marked for deletion, retrying in 5 seconds...')
                    time.sleep(5)
                else:
                    raise e

        table_client = TableClient.from_connection_string(conn_str=connection_string, table_name=table_name)
        json_file_path = Path(__file__).resolve().parent / "db" / f"{table_name}.json"
        with open(json_file_path, "r", encoding="utf8") as file:
            entities = json.load(file)

        for entity in entities["rows"]:
            row_key = str(entity.get("id", uuid.uuid4()))
            # Convert any nested objects to JSON strings
            for key, value in entity.items():
                if isinstance(value, (dict, list)):
                    entity[key] = json.dumps(value)
            
            table_client.create_entity({
                'PartitionKey': table_name,
                'RowKey': row_key,
                **entity
            })
            print(f"Added entity to {table_name} with key {row_key}")

def main(connection_string="UseDevelopmentStorage=true", reset=False):
    table_service_client = TableServiceClient.from_connection_string(conn_str=connection_string)

    if reset:
        delete_tables(table_service_client, connection_string)

    create_and_populate_tables(table_service_client, connection_string)

if __name__ == "__main__":
    main()

File: samples/cext-trey-research-csharp/DbSetup/AzureTableSetup.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;
using Azure;
using Azure.Data.Tables;

namespace Trey.Research.Declarative.Agent.DbSetup
{
    public class AzureTableSetup
    {
        private readonly string _connectionString;
        private readonly string[] TableNames = { "Project", "Consultant", "Assignment" };

        public AzureTableSetup(string connectionString)
        {
            _connectionString = connectionString;
        }

        public async Task SetupTablesAndDataAsync(bool reset = true)
        {
            var tableServiceClient = new TableServiceClient(_connectionString);

            if (reset)
            {
                await DeleteTablesAsync(tableServiceClient);
            }

            foreach (var tableName in TableNames)
            {
                await CreateAndPopulateTableAsync(tableServiceClient, tableName);
            }
        }

        private async Task DeleteTablesAsync(TableServiceClient tableServiceClient)
        {
            foreach (var tableName in TableNames)
            {
                try
                {
                    var tableClient = tableServiceClient.GetTableClient(tableName);
                    Console.WriteLine($"Deleting table: {tableName}");
                    await tableClient.DeleteAsync();
                }
                catch (RequestFailedException ex) when (ex.Status == 404)
                {
                    Console.WriteLine($"Table {tableName} does not exist.");
                }
                catch (RequestFailedException ex)
                {
                    Console.WriteLine($"Error deleting table {tableName}: {ex.Message}");
                }
            }

            Console.WriteLine("Waiting for tables to be deleted...");
            await Task.Delay(1000); // Allow some time for table deletion
        }

        private async Task CreateAndPopulateTableAsync(TableServiceClient tableServiceClient, string tableName)
        {
            try
            {
                var tableClient = tableServiceClient.GetTableClient(tableName);
                Console.WriteLine($"Creating table: {tableName}");

                // Create table if it does not exist
                try
                {
                    await tableClient.CreateAsync();
                }
                catch (RequestFailedException ex) when (ex.Status == 409)
                {
                    Console.WriteLine($"Table {tableName} already exists.");
                }
                // Populate the table
                var jsonString = File.ReadAllText($"DbSetup/db/{tableName}.json");
                var entities = JsonSerializer.Deserialize<JsonDocument>(jsonString);

                if (entities?.RootElement.TryGetProperty("Rows", out var rows) == true)
                {
                    foreach (var row in rows.EnumerateArray())
                    {
                        var rowKey = row.GetProperty("Id").GetString() ?? Guid.NewGuid().ToString();
                        var entity = new Dictionary<string, string>();

                        foreach (var prop in row.EnumerateObject())
                        {
                            // Store complex types as JSON strings
                            if (prop.Value.ValueKind == JsonValueKind.Object || prop.Value.ValueKind == JsonValueKind.Array)
                            {
                                entity[prop.Name+"Json"] = prop.Value.GetRawText(); // Store as JSON string
                            }
                            else
                            {
                                entity[prop.Name] = ConvertToString(prop.Value);
                            }
                        }

                        var tableEntity = new TableEntity(tableName, rowKey)
                        {
                            { "partitionKey", tableName },
                            { "rowKey", rowKey }
                        };

                        foreach (var kvp in entity)
                        {
                            tableEntity[kvp.Key] = kvp.Value;
                        }

                        await tableClient.AddEntityAsync(tableEntity);
                        Console.WriteLine($"Added entity to {tableName} with key {rowKey}");
                    }
                }
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Error creating or populating table {tableName}: {ex.Message}");
            }
        }

        private string ConvertToString(JsonElement element)
        {
            return element.ValueKind switch
            {
                JsonValueKind.String => element.GetString(),
                JsonValueKind.Number => element.GetDecimal().ToString(),
                JsonValueKind.True => "True",
                JsonValueKind.False => "False",
                JsonValueKind.Object or JsonValueKind.Array => element.ToString(),
                _ => throw new InvalidOperationException("Unsupported JSON value kind.")
            };
        }
    }
}